### ISSUE TITLE
I don't friggin care that template studio loaded a config file

### ISSUE THREAD

**Issue by amoldeshpande** [Microsoft/WindowsTemplateStudio#917]
```
Windows Template Studio 2017-08-08 20:43:06.532
Version: 1.2.17198.1
>
[20:43:06.282 - Information]::Configuration file loaded: F:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\Extensions\uguvjkvi.scm\WindowsTemplateStudio.config.json
```
### Repro steps
Open a Solution

### Expected Behavior

Should not see any WTS output unless there is an error.

### Actual Behavior
Every time I open a solution, the output window switches to Windows Template Studio  with the above text.

I really don't understand why every extension  thinks it's the most important  snowflake in the world.
Please stop hijacking my output window.

**Comment by mrlacey** [comment]
_(I know this is long but bear with me as I have a couple of solution proposals at the end.)_

There are actually a few things going on here. 

1. When VS opens a solution (new or existing) the WTS extension is loaded.
2. The WTS extension checks for new templates. (This should make creating a new project with WTS or adding to an existing one faster.)
3. As part of the process of checking for new templates, it outputs some debug information.
4. The debug information is put in a separate pane, to make it easy to isolate and identify. 
5. VS switches to different output panes based on factors I can't explicitly nail down but are either due to the order of the pane's creation or which ever was written to last.

In reverse order.
5) We can't do anything about this ourselves. It's what VS does. I've also seen VS automatically switch to some obscure other panes for things I didn;t even know I had installed. I've also seen other reports of people complaining about other extensions showing content in the output pane when not directly related to the solution that was opened. I don't think this is a WTS issue or anything we can alter beyond feeding back to the VS team to see if anything can be done about controlling which output pane is shown.
4) Having all the WTS related information in a single pane is really helpful when there are issues. I can't think of a good reason for changing this.
3) We definitely could do something here. **Would having the ability to control how much output is shown be a suitable solution here?** This would mean adding a setting to limit what output is shown. You could then change it to, for example, only show errors. If there were no errors, there would be nothing to show and the pane would not be created. (It's created when we first have something to show although this may need to change slightly if we added this feature.) This would default to the current behavior of showing everything.
2) Automatically checking for new templates could be made optional.. **Would having control over whether the extension automatically checks for new templates in the background be useful?** We could add a setting to turn this off (it would stay on by default.) Disabling this could also help those who have a very slow solution load experience by having VS do less. Probably not a major impact but some people may get a perf improvement by disabling this.
1 ) This is just the way VS handles extension loading. Nothing we can do here.

So, in summary, I can see two things we can (and might make sense to) change.
- Allow control over what level of output is shown.
- Allow control over whether the extension automatically tries to proactively check for new templates when loaded.

We can do either or both of these. Thoughts and feedback, please.

**Comment by amoldeshpande** [comment]
Thanks for the detailed response. My thoughts:

Option 1 : A knob to turn up or down the logging would be excellent. That way we could debug it when needed and avoid knowing about WTS otherwise. Which is how life should work :)

Option 2:  Controlling when to check for new templates, especially if it helps with loading times, is also a great feature.  Is it possible for the extension to detect when a user is creating a new project ? searching for templates at that point seems more optimal than solution load.  

Visual Studio drives me crazy with how much extensions affect the user experience and anything an extension can do to stay out of the way would be much appreciated.

thanks!

**Comment by mrlacey** [comment]
When I wrote the two options I felt they were really both needed and so will look at adding them. 
Will update the title of this issue to reflect what the plan is to address this.
Shared dependent code (#919) and other priorities mean this will won't make this release but is definitely on my schedule.

**Changes Needed**

1. Ability to control what level of information is output via the ShellHealthWriter. 
Three options:
Errors: includes just Critical and Error events
Info: includes above plus Information & Warning events
Verbose: includes everything (default option.)
Output pane should not be created until there's something to show based on above setting

2. Add a configurable setting to control whether the extension will try and check for new templates when first loaded. This must default to 'on' meaning it will check. If enabled the current behavior will remain but if disabled the extension will only check for updates when the wizard is launched.

Depends on new extension settings support. (Coming with #919)

**Comment by amoldeshpande** [comment]
I question the decision to keep the check during solution load. The frequency of creating a WTS project is much lower than opening an arbitrary solution with existing projects. 

Please do not make the 90% case suffer for a pretty marginal optimization

**Comment by mrlacey** [comment]
My concern is with changing the current behavior. Especially without fully investigating all the consequences.
I also don't think there's really a lot of suffering. I know you have a strong dislike for what happens in the output pane but beyond that we're really just talking about a single HTTP request.

**Comment by amoldeshpande** [comment]
Well you're solving the output window issue with fix 1. My concern now is startup time for visual studio to actually get to my work. From my memory of the last bug i filed, the templates packages are brought down initially from the wizard anyway. 

Suffering is an addictive function of all the extensions that each think they're not creating suffering :)

**Comment by mrlacey** [comment]
It'd be good to get wider feedback on what people think about the default behavior for automatic checking for new templates. Should it be enabled by default?
If we can avoid showing up in the slow extension prompt that can only be a good thing.
Another fact to consider here is how often templates actually change. (hint. it's not very often.)

**Comment by amoldeshpande** [comment]
ok, fair enough to get more feedback.

One last point I want  to make is that if every extension I load is making a single HTTP call, that could potentially be a lousy startup experience if I'm offline or having network problems where it takes 30 seconds for each request to time out (default).

**Comment by crutkas** [comment]
@amoldeshpande are you implying the HTTP call is causing VS to be slower at startup?  It shouldn't

**Comment by amoldeshpande** [comment]
I'm basing that comment on how something like Perforce behaves when it can't contact a server. 

If the WTS check is purely asynchronous then I suppose I don't really care (except from a purely design POV) if it does that  on load. 

My main concern is  that WTS not add any more to my solution load time than all the other extensions that I'm also forced to load do.  If y'all think you can do that with your proposed designs, then consider me satisfied.

**Comment by crutkas** [comment]
What you have there, it only occurs when you start the WTS project.  That info literally says "we loaded a file".  Not downloading a file.

Do you feel WTS added load time?  I'm 99% sure we don't do anything at VS start up else we would show up in the VS performance window. (Help->manage VS performance)

**Comment by mrlacey** [comment]
What the extension does on startup is in a background thread and should only have an impact on startup if you have a very slow/underpowered machine. If that's the case then you'll likely have more issues than what the extension does on load.

**Comment by amoldeshpande** [comment]
>What you have there, it only occurs when you start the WTS project. That info literally says "we loaded a file". Not downloading a file.

No, it occurs every time I load a solution that has nothing to do with WTS.  Mark has already addressed that so I really don't want to beat that dead horse.

The only thing I said about loading time was:

>Option 2: Controlling when to check for new templates, especially if it helps with loading times

If it makes no difference, I don't really care. 
I am not here to argue the philosophy behind why an extension that is not involved in any project in my solution should  even be running.

**Comment by mrlacey** [comment]
Sorry @crutkas there's more happening at startup than you realize.

I just went offline and created a new console app. The output to the WTS pane was this.


> Windows Template Studio 2017-08-10 15:59:24.804
> Version: 0.14.17222.65001
> 
> [15:59:23.848 - Verbose]::Package certificate OID.0.9.2342.19200300.100.1.1=10018646, CN=Microsoft Windows App Studio, OU=Microsoft Windows App Studio
> [15:59:24.912 - Information]::Configuration file loaded: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Common7\IDE\Extensions\uguvjkvi.scm\WindowsTemplateStudio.config.json
> [15:59:24.984 - Information]::Configuration file loaded: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Common7\IDE\Extensions\15yx3tlo.1aj\WindowsTemplateStudio.config.json
> [15:59:24.918 - Information]::Configuration file loaded: C:\Program Files (x86)\Microsoft Visual Studio\Preview\Enterprise\Common7\IDE\Extensions\jlr0zr1z.nt1\WindowsTemplateStudio.config.json
> [15:59:37.281 - Verbose]::Templates extracted to C:\Users\matt\AppData\Local\Temp\0fxv33yy.eb4.
> [15:59:37.983 - Information]::Templates content can't be downloaded right now, we will try it later.
> [15:59:38.004 - Error]::Error downloading from https://wtsrepository.blob.core.windows.net/dev/Latest/Templates.mstx. Internet connection is required to download template updates.
> ----------- Addtional Exception Info -----------
> System.Net.WebException: The remote name could not be resolved: 'wtsrepository.blob.core.windows.net'
>    at System.Net.WebClient.DownloadFile(Uri address, String fileName)
>    at System.Net.WebClient.DownloadFile(String address, String fileName)
>    at Microsoft.Templates.Core.Locations.RemoteTemplatesSource.DownloadContent(String sourceUrl, String file)
> -----------------------------------------------
> [15:59:40.365 - Information]::Templates content can't be downloaded right now, we will try it later.
> [15:59:40.373 - Error]::Error downloading from https://wtsrepository.blob.core.windows.net/pre/Latest/Templates.mstx. Internet connection is required to download template updates.
> ----------- Addtional Exception Info -----------
> System.Net.WebException: The remote name could not be resolved: 'wtsrepository.blob.core.windows.net'
>    at System.Net.WebClient.DownloadFile(Uri address, String fileName)
>    at System.Net.WebClient.DownloadFile(String address, String fileName)
>    at Microsoft.Templates.Core.Locations.RemoteTemplatesSource.DownloadContent(String sourceUrl, String file)
> -----------------------------------------------
> [15:59:40.422 - Information]::Templates content can't be downloaded right now, we will try it later.
> [15:59:40.435 - Error]::Error downloading from https://wts.azureedge.net/Templates.mstx. Internet connection is required to download template updates.
> ----------- Addtional Exception Info -----------
> System.Net.WebException: The remote name could not be resolved: 'wts.azureedge.net'
>    at System.Net.WebClient.DownloadFile(Uri address, String fileName)
>    at System.Net.WebClient.DownloadFile(String address, String fileName)
>    at Microsoft.Templates.Core.Locations.RemoteTemplatesSource.DownloadContent(String sourceUrl, String file)
> -----------------------------------------------


There's something definitely not right here. 
It's clearly doing more than just loading a config file.
It's loading three of them because I have three versions of the extension installed (because dev) and then for each, it's trying to download a file. As evidenced by the three exceptions (because offline.)

Checking the code, the extension will always forcibly download the latest version of the templates and ignore the cache TTL value (VersionCheckingExpirationMinutes) which kind of makes that value pointless. This is definitely a bug but this whole thing needs revisiting.

**Comment by sibille** [comment]
Matt, it should ignore the VersionCheckingExpirationMinutes value for the local template source only, not dev, pre and pro). I've currently installed only the pro version, and it's not trying to download anything when I'm offline, but I'll double check tomorrow. Maybe a side-by-side issue?

**Comment by sibille** [comment]
It's not a side by side issue, happening on my machine too, the first time you create/load a solution on a new VS Instance the extension is checking for updates. Will look into this.

**Comment by ralarcon** [comment]
The extension (up to version 1.2) was automatically  intitialized in background as Visual Studio was launched. This will not happen anymore as well as the templates updates verification / downloaded will not be launched util the WTS wizard demand it thru new project or right-click (no more diffenent versions checking for templates updates concurrently). This is related to issue #928 and fixed in the PR #1088.

Regarding the trace leve options by default, the pre / dev versions have more "verbosity" than production version. The trace level is currently configurable (in the json config file, there is no friendly UI to do so).

If you agree @mrlacey, @crutkas and @amoldeshpande we can close this issue.

**Comment by mrlacey** [comment]
@ralarcon is the trace level setting (and anything else in it) documented anywhere?

**Comment by ralarcon** [comment]
My hope was to be self-descriptive / documented. It is in the [Configuration](https://github.com/Microsoft/WindowsTemplateStudio/blob/master/code/src/Core/Configuration/Configuration.cs) class and the [WindowsTemplateStudio.tokenized.config.json](https://github.com/Microsoft/WindowsTemplateStudio/blob/master/code/src/Core/Configuration/WindowsTemplateStudio.tokenized.config.json), the `DiagnosticsTraceLevel` property.

**Comment by mrlacey** [comment]
The request here is to have a way to configure something.
You say there's a way to do this that you hoped it was "self-descriptive / documented".
This isn't the case.

If you don't want (or think it's worth having) it in some settings UI, that supposes that:
1. people know that there is a configuration file that exists which controls this 
2. that people will know where to find it 
3. what values can be used. 

I don't think any of those assumptions can be made.

- People won't know what's possible unless they are told. At a minimum add this to [the getting started doc](https://github.com/Microsoft/WindowsTemplateStudio/blob/master/docs/getting-started-extension.md).
- The documentation should tell them where they can find the config file on their machine.
- The documentation should list the values that can be set and the effect each will have.

**Comment by ralarcon** [comment]
I was answering you regarding the TracelLevel thing... You are right, the configuration documentation is not available as a public documentation. You only have access thru reading the code. 

We can add a section explaining the configuration file and how the configuration is applied. By the way, we can add some kind of page in the wizard to handle the configuration but the initiall thought was to include the configuration UI thru VS Options (more complex approach, by the way)

**Comment by mrlacey** [comment]
This whole issue is only open because I'm adding other VS settings and so adding this as a configurable option was a small addition. (Hence this having a dependency on #919)
`DiagnosticsTraceLevel` controls all trace output. This issue was to only to control what was displayed in the output window. This could leave full logging to file. etc. so any issues could be investigated more thoroughly if needed.

I suggest:
- Make this issue only about configuring what goes to the output window
- Ignore the proposed setting for when checks are made based on #928
- Add new documentation on how to change the config setting manually as a short-term workaround.


Also bumped this to 1.5 as won't happen for 1.4 if that's going to be a small, quick release.
